dali/kernels/transpose/transpose_gpu_impl_test.cu

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/kernels/transpose/transpose_gpu_impl.cuh"   // NOLINT
+#include "dali/kernels/transpose/transpose_gpu_setup.cuh"  // NOLINT
+#include <gtest/gtest.h>
+#include <algorithm>
+#include <numeric>
+#include <vector>
+#include "dali/core/dev_buffer.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/core/tensor_shape_print.h"
+#include "dali/test/test_tensors.h"
+#include "dali/core/cuda_event.h"
+#include "dali/kernels/transpose/transpose_test.h"
+
+namespace dali {
+namespace kernels {
+
+using namespace transpose_impl;  // NOLINT
+
+
+TEST(SimplifyPermute, NoSimplification) {
+  int64_t shape[] = { 2, 3, 4, 5 };
+  int perm[] = { 0, 3, 2, 1 };
+  TensorShape<> s_shape, ref_shape;
+  SmallVector<int, 6> s_perm, ref_perm;
+  SimplifyPermute(s_shape, s_perm, shape, perm, 4);
+  ref_shape = { 2, 3, 4, 5 };
+  ref_perm = { 0, 3, 2, 1 };
+  EXPECT_EQ(s_shape, ref_shape);
+  EXPECT_EQ(s_perm, ref_perm);
+}
+
+TEST(SimplifyPermute, CollapseUnitDims) {
+  int64_t shape[] = { 2, 1, 3, 4, 1, 5 };
+  int perm[] = { 0, 5, 1, 3, 2, 4 };
+  TensorShape<> s_shape, ref_shape;
+  SmallVector<int, 6> s_perm, ref_perm;
+  SimplifyPermute(s_shape, s_perm, shape, perm, 6);
+  ref_shape = { 2, 3, 4, 5 };
+  ref_perm = { 0, 3, 2, 1 };
+  EXPECT_EQ(s_shape, ref_shape);
+  EXPECT_EQ(s_perm, ref_perm);
+}
+
+TEST(SimplifyPermute, Collapse) {
+  int64_t shape[] = { 2, 1, 3, 4, 1, 5 };
+  int perm[] = { 3, 4, 5, 0, 1, 2 };
+  TensorShape<> s_shape, ref_shape;
+  SmallVector<int, 6> s_perm, ref_perm;
+  SimplifyPermute(s_shape, s_perm, shape, perm, 6);
+  ref_shape = { 6, 20 };
+  ref_perm = { 1, 0 };
+  EXPECT_EQ(s_shape, ref_shape);
+  EXPECT_EQ(s_perm, ref_perm);
+}
+
+TEST(TransposeTiled, AllPerm4DInnermost) {
```

Comment:
Can you test `GetTransposeMethod` to make sure the right strategy is selected?

Reply:
In principle, yes.

Reply:
So please add if it is not a huge effort.

Reply:
Done.